File: 133-clone-graph/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
    unordered_map<int, Node*> mp;
public:
    Node* cloneGraph(Node* node) {
        if(node == NULL) return NULL;
        Node* clone = new Node();
        clone -> val = node -> val;
        mp[node -> val] = clone;
        for(Node* nn : node -> neighbors) {
            if(mp.find(nn -> val) != mp.end())  (clone -> neighbors).push_back(mp[nn -> val]);
            else {
                Node *n = cloneGraph(nn);
                (clone -> neighbors).push_back(n);
                mp[nn -> val] = n;
            }
        }
        return clone;
    }
};

File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        HashMap<Integer, Node> visited = new HashMap<>();
        Node clone = new Node(node.val);
        Queue<Node> q1 = new LinkedList<>(), q2 = new LinkedList<>();
        q1.add(node);
        q2.add(clone);
        visited.put(clone.val, clone);
        while(q1.isEmpty() == false && q2.isEmpty() == false) {
            Node cur = q1.remove();
            Node curclone = q2.remove();
            for(Node n : cur.neighbors) {
                Node insertingNode;
                if(visited.get(n.val) == null) {
                    insertingNode = new Node(n.val);
                    q1.add(n);
                    q2.add(insertingNode);
                    visited.put(n.val, insertingNode);
                } else insertingNode = visited.get(n.val);
                curclone.neighbors.add(insertingNode);
            }
        }
        return clone;
    }
}